tensorflow_probability/python/experimental/mcmc/covariance_reducer.py

```diff
@@ -0,0 +1,260 @@
+    - `event_ndims=2` treats the samples as a single random variable of
+      shape `[5, 7]` and computes its covariance.  The shape of the result
+      is `[5, 7, 5, 7]`.
+
+    For nested latent samples, both `event_ndims` and `ddof` must be either
+    a single value or an identical structure. For example, if the chain state
+    is the tuple of dictionaries ({pairA, pairB}, {pairC}), one could use
+    scalar values of `event_ndims` (integer or `None`) and `ddof` (integer) that
+    apply to all states in the latent. If not, `event_ndims` and `ddof` must
+    have structures that ressemble ({keyA: argA, keyB: argB}, {keyC: argC}).
+
+    Args:
+      event_ndims: A (possibly nested) structure of integers, or `None`. Defines
+        the number of inner-most dimensions that represent the event shape.
+        Specifying `None` returns all cross product terms (no batching)
+        and is the default.
+      ddof: A (possibly nested) structure of integers that represent the
+        requested dynamic degrees of freedom. For example, use `ddof=0`
+        for population covariance and `ddof=1` for sample covariance. Defaults
+        to the population covariance.
+      name: Python `str` name prefixed to Ops created by this function.
+        Default value: `None` (i.e., 'covariance_reducer').
+    """
+    self._parameters = dict(
+        event_ndims=event_ndims,
+        ddof=ddof,
+        name=name or 'covariance_reducer'
+    )
+
+  def initialize(self, initial_chain_state=None, initial_kernel_results=None):
+    """Initializes a `RunningCovarianceState` using previously defined metadata.
+
+    For calculation purposes, the `initial_chain_state` does not count as a
+    sample. This is a deliberate decision that ensures consistency across
+    sampling procedures (i.e. `tfp.mcmc.sample_chain` follows similar
+    semantics).
+
+    Args:
+      initial_chain_state: A (possibly nested) structure of `Tensor`s or Python
+        `list`s of `Tensor`s representing the current state(s) of the Markov
+        chain(s). It is used to infer the shape and dtype of future samples.
+      initial_kernel_results: A (possibly nested) `tuple`, `namedtuple` or
+        `list` of `Tensor`s representing internal calculations made in a related
+        `TransitionKernel`. For streaming covariance, this argument also has no
+        influence on computation; hence, it is also `None` by default.
+        Nonetheless, this argument is still accepted as it will be supplied by
+        the Streaming MCMC framework.
+
+    Returns:
+      state: `RunningCovarianceState` representing a stream of no inputs and
+        with identical structure to the `initial_chain_state`.
+    """
+    with tf.name_scope(
+        mcmc_util.make_name(self.name, 'covariance_reducer', 'initialize')):
+      initial_chain_state = tf.nest.map_structure(
+          tf.convert_to_tensor,
+          initial_chain_state)
+      self._parameters['shape'] = tf.nest.map_structure(
```

Comment:
This doesn't feel right, why are we mutating the class state in a method? In TFP we are *very* careful not to use mutable state, only violating this directive for caching purposes (and internal mutabiity of Variables).

I think the right solution here is to re-compute all these properties in initialize/one_step/finalize, creating new `RunningCovariance` objects each time.

Either way, storing these side-effects in _parameters is incorrect as well, as they're not parameters (which are exclusively understood to be the arguments to the class initializer in TFP).

Reply:
See _prepare_args pattern in a few places in TFP for what I mean.

Reply:
Yeah, I was wondering what you would think of the mutation.

For my part, I had let it slide because the expected use pattern leads to the mutation occurring outside the sampler's `while_loop`, but you're right, `prepare_args` is probably better.